### PR TITLE
Add Sort feature

### LINKED ISSUE #47: Add Sort Feature
Add Sort command that can sort listings in a certain order

- Alphabetical
- Deadline
- Created Date
(Can choose ascending/descending order)

### LINKED ISSUE #48: As a user, I can sort the projects in different orders
so that I can view projects in a more convenient way according to my needs.

### PR DESCRIPTION
**Pull request by kkangs0226**
List of projects displayed cannot be sorted in a specified order.

Adding sort feature will make application more user friendly as it
can display projects in orders required by users.

Let's
* add SortCommand and SortCommandParser
* add other relevant Comparators for sorting
* make necessary changes to existing files
* add tests for SortCommand, SortCommandParser and Comparators

Closes #47, closes #48

### CODE REVIEW

File: src/main/java/seedu/momentum/logic/commands/SortCommand.java
Comment (khoodehui) [inline src/main/java/seedu/momentum/logic/commands/SortCommand.java:53]: Newline before param

File: src/main/java/seedu/momentum/logic/parser/SortCommandParser.java
Comment (khoodehui) [inline src/main/java/seedu/momentum/logic/parser/SortCommandParser.java:21]: Newline after description

File: src/main/java/seedu/momentum/logic/commands/SortCommand.java
Comment (pr4aveen) [inline src/main/java/seedu/momentum/logic/commands/SortCommand.java:44]: Should this go under `seedu.momentum.commons.core.Messages`?

File: src/main/java/seedu/momentum/model/ModelManager.java
Comment (pr4aveen) [inline src/main/java/seedu/momentum/model/ModelManager.java]: Should these 3 variables just be declared directly since their values aren't passed into the constructor?

File: src/main/java/seedu/momentum/model/project/comparators/CreatedDateCompare.java
Comment (pr4aveen) [inline src/main/java/seedu/momentum/model/project/comparators/CreatedDateCompare.java:30]: Should we implement a `compareTo` method in `Date.java`?

File: src/main/java/seedu/momentum/model/project/comparators/DeadlineCompare.java
Comment (pr4aveen) [inline src/main/java/seedu/momentum/model/project/comparators/DeadlineCompare.java]: Incomplete header here

File: src/main/java/seedu/momentum/model/project/comparators/DeadlineCompare.java
Comment (pr4aveen) [inline src/main/java/seedu/momentum/model/project/comparators/DeadlineCompare.java]: I think we should have `compareTo` methods in the `Date` and `Time` classes and use them here instead

File: src/main/java/seedu/momentum/model/project/Project.java
Comment (pr4aveen) [inline src/main/java/seedu/momentum/model/project/Project.java]: I'm not sure if there's a cleaner way to do this. What do the rest think?

File: src/main/java/seedu/momentum/model/project/comparators/DeadlineCompare.java
Comment (pr4aveen) [inline src/main/java/seedu/momentum/model/project/comparators/DeadlineCompare.java]: This part is related to the comment on `Project.java`

File: src/test/java/seedu/momentum/logic/commands/SortCommandTest.java
Comment (pr4aveen) [inline src/test/java/seedu/momentum/logic/commands/SortCommandTest.java]: I think you can extract the common string variables that you have reused in this section and make them private static variables. For example, `ascending`, `descending`, `alphabetical`, `deadline`, `created date`

File: src/main/java/seedu/momentum/logic/parser/SortCommandParser.java
Comment (pr4aveen) [inline src/main/java/seedu/momentum/logic/parser/SortCommandParser.java]: You might want to make these public static variables instead. It's used quite a fair bit in tests and it'll be better if they all refer to one variable

File: src/test/java/seedu/momentum/logic/parser/ProjectBookParserTest.java
Comment (pr4aveen) [inline src/test/java/seedu/momentum/logic/parser/ProjectBookParserTest.java]: Might want to replace the `alpha` (and other keywords in other classes) with the public static variables I recommended in `SortCommandParser.java`

File: src/main/java/seedu/momentum/logic/parser/SortCommandParser.java
Comment (pr4aveen) [inline src/main/java/seedu/momentum/logic/parser/SortCommandParser.java:25]: Just checking, what's the expected behaviour when there is a preamble?

For example, `sort hello type/alpha order/dsc`

File: src/test/java/seedu/momentum/logic/commands/AddCommandIntegrationTest.java
Comment (pr4aveen) [inline src/test/java/seedu/momentum/logic/commands/AddCommandIntegrationTest.java]: Is `Hello` referring to `Dana`?

**Review by pr4aveen — APPROVED** [review]
Have a few comments that might need to be addressed before merging. Should be ok to merge once those changes have been made :)

File: src/main/java/seedu/momentum/logic/parser/SortCommandParser.java
Comment (khoodehui) [inline src/main/java/seedu/momentum/logic/parser/SortCommandParser.java]: Code might be easier to understand if `parseSortOrder()` only returns the argument which the user entered in the command. Then under the variable declaration for `isAscending()`, do an `equal()` check to see whether the argument the user provided is equal to `ASCENDING_ORDER`.

This is so that the nature of the `parseSortOrder()` method is also consistent with that of the `parseSortType()` method.

File: src/main/java/seedu/momentum/model/project/comparators/CreatedDateCompare.java
Comment (khoodehui) [reply src/main/java/seedu/momentum/model/project/comparators/CreatedDateCompare.java:30]: I think we should.

File: src/test/java/seedu/momentum/logic/parser/SortCommandParserTest.java
Comment (khoodehui) [inline src/test/java/seedu/momentum/logic/parser/SortCommandParserTest.java]: Noticed that you have declared various form of `SortCommands` as private static final variables in `SortCommandTest.java`, and they are repeated here. Will it be better to just abstract all of the various `SortCommands` into a utility class, then have the test classes get the different commands from there.

**Review by khoodehui — APPROVED** [review]
Just some things to take note, not too much of an issue though. Otherwise LGTM!

File: src/test/java/seedu/momentum/logic/commands/AddCommandIntegrationTest.java
Comment (kkangs0226) [reply src/test/java/seedu/momentum/logic/commands/AddCommandIntegrationTest.java]: Yes I'll go fix it!

File: src/main/java/seedu/momentum/model/project/comparators/DeadlineCompare.java
Comment (kkangs0226) [reply src/main/java/seedu/momentum/model/project/comparators/DeadlineCompare.java]: Fixed!

File: src/main/java/seedu/momentum/model/Model.java
Comment (claracheong4) [inline src/main/java/seedu/momentum/model/Model.java:92]: New line needed before @throws

File: src/main/java/seedu/momentum/model/project/UniqueProjectList.java
Comment (claracheong4) [inline src/main/java/seedu/momentum/model/project/UniqueProjectList.java]: Make name implement a comparable instead?

File: src/main/java/seedu/momentum/model/project/UniqueProjectList.java
Comment (claracheong4) [inline src/main/java/seedu/momentum/model/project/UniqueProjectList.java]: The method is quite long, maybe extract this into another method?

File: src/main/java/seedu/momentum/model/project/UniqueProjectList.java
Comment (claracheong4) [inline src/main/java/seedu/momentum/model/project/UniqueProjectList.java]: similarly for this

File: src/main/java/seedu/momentum/model/project/UniqueProjectList.java
Comment (claracheong4) [inline src/main/java/seedu/momentum/model/project/UniqueProjectList.java]: and this?

**Review by claracheong4 — COMMENTED** [review]
I think some minor changes has to be made.

File: src/main/java/seedu/momentum/logic/parser/SortCommandParser.java
Comment (kkangs0226) [reply src/main/java/seedu/momentum/logic/parser/SortCommandParser.java:25]: It should disregard the preamble. 
I have added a test for this!

File: src/main/java/seedu/momentum/logic/commands/SortCommand.java
Comment (kkangs0226) [reply src/main/java/seedu/momentum/logic/commands/SortCommand.java:53]: Somehow the checkstyle doesn't pass it if i leave a newline, I think we can disregard these for now?

File: src/main/java/seedu/momentum/logic/parser/SortCommandParser.java
Comment (kkangs0226) [reply src/main/java/seedu/momentum/logic/parser/SortCommandParser.java:21]: Same as my previous comment

File: src/main/java/seedu/momentum/logic/commands/SortCommand.java
Comment (kkangs0226) [reply src/main/java/seedu/momentum/logic/commands/SortCommand.java:44]: I think we can leave it here since other classes are doing the same. We can move them to Messages in one go in the next iteration!

File: src/main/java/seedu/momentum/model/ModelManager.java
Comment (kkangs0226) [reply src/main/java/seedu/momentum/model/ModelManager.java]: Fixed!

File: src/main/java/seedu/momentum/model/project/Project.java
Comment (kkangs0226) [reply src/main/java/seedu/momentum/model/project/Project.java]: The reason why I need to use HashMap is because not all projects have Deadlines, and I'm using Comparator.nullsLast to place all projects without deadlines at the end of the list after projects with deadlines, so I need to compare Deadlines (and not Projects). But after extracting the deadlines (or nulls), I will still need to compare the names alphabetically in case of same deadline or null deadline, which is why I need to return the name as well. Anyone can lmk if there is a cleaner way to do this!

File: src/test/java/seedu/momentum/logic/commands/SortCommandTest.java
Comment (kkangs0226) [reply src/test/java/seedu/momentum/logic/commands/SortCommandTest.java]: Fixed!

File: src/main/java/seedu/momentum/logic/parser/SortCommandParser.java
Comment (kkangs0226) [reply src/main/java/seedu/momentum/logic/parser/SortCommandParser.java]: Fixed!

File: src/test/java/seedu/momentum/logic/parser/ProjectBookParserTest.java
Comment (kkangs0226) [reply src/test/java/seedu/momentum/logic/parser/ProjectBookParserTest.java]: Fixed!

File: src/main/java/seedu/momentum/logic/parser/SortCommandParser.java
Comment (kkangs0226) [reply src/main/java/seedu/momentum/logic/parser/SortCommandParser.java:25]: I have changed the SortCommandParser to fail when there is a non-empty preamble for consistency as AddCommandParser and FindCommandParser does the same. I have added a test accordingly!

File: src/main/java/seedu/momentum/logic/parser/SortCommandParser.java
Comment (kkangs0226) [reply src/main/java/seedu/momentum/logic/parser/SortCommandParser.java]: Fixed!

File: src/test/java/seedu/momentum/logic/parser/SortCommandParserTest.java
Comment (kkangs0226) [reply src/test/java/seedu/momentum/logic/parser/SortCommandParserTest.java]: Abstracted into a SortCommandUtil class!

File: src/main/java/seedu/momentum/model/Model.java
Comment (kkangs0226) [reply src/main/java/seedu/momentum/model/Model.java:92]: Fixed!

File: src/main/java/seedu/momentum/model/project/UniqueProjectList.java
Comment (kkangs0226) [reply src/main/java/seedu/momentum/model/project/UniqueProjectList.java]: Fixed!

File: src/main/java/seedu/momentum/model/project/UniqueProjectList.java
Comment (kkangs0226) [reply src/main/java/seedu/momentum/model/project/UniqueProjectList.java]: Fixed!

File: src/main/java/seedu/momentum/model/project/UniqueProjectList.java
Comment (kkangs0226) [reply src/main/java/seedu/momentum/model/project/UniqueProjectList.java]: Fixed!

File: src/main/java/seedu/momentum/model/project/comparators/CreatedDateCompare.java
Comment (kkangs0226) [reply src/main/java/seedu/momentum/model/project/comparators/CreatedDateCompare.java:30]: Done!

File: src/main/java/seedu/momentum/model/project/comparators/DeadlineCompare.java
Comment (kkangs0226) [reply src/main/java/seedu/momentum/model/project/comparators/DeadlineCompare.java]: Done as well!

File: src/main/java/seedu/momentum/model/project/UniqueProjectList.java
Comment (kkangs0226) [reply src/main/java/seedu/momentum/model/project/UniqueProjectList.java]: Done!

**Pull request merged**
